---
title: "assignment.rmd-week03"
author: "Robin Scheuplein"
date: "`r Sys.Date()`"
output: html_document
---

```{r setup}
suppressPackageStartupMessages({
  library(AnnotationHub)  #retrieve genomes/annotations
  library(Rsubread) #alignment
  library(rtracklayer) #import/export files
  library(Biostrings)
  library(Rfastp) #adapter tripping, QC of reads
  library(epiwraps) #visualiztation
  library(Rsamtools)
library(GenomicAlignments)})
```

# Downloading the raw reads

```{r, eval=FALSE}
#download raw reads
options(timeout=3600) # we need to increase the download timeout since the file is large and cnx slow
dir.create("raw") #create new folder
url <- "https://www.encodeproject.org/files/ENCFF127RRR/@@download/ENCFF127RRR.fastq.gz"
destination <- "raw/ENCFF127RRR.fastq.gz"
download.file(url, destfile = destination, mode = "wb") #download FASTQ file
```

# Reads QC and trimming

## Using R and Rfastp

```{r}
# trim away adaptors
dir.create("rfastp.trimmed")
qc <- lapply(c(ENCFF127RRR="raw/ENCFF127RRR.fastq.gz"), FUN=function(x){
  Rfastp::rfastp(x, thread=4, overrepresentationAnalysis=TRUE,
                 outputFastq=file.path("rfastp.trimmed/",gsub("\\.fastq\\.gz$","",basename(x))))
})
```

```{r, eval=FALSE}
Rfastp::curvePlot(qc$ENCFF127RRR, curve="content_curves")
```

# Alignment

## Using Rsubread

### Building a genome index for mapping

```{r}
# Search and download drosophila melanogaster genome sequence (dna_sm) in TwoBit/2bit format for BDGP6
ah <- AnnotationHub() #load AnnotationHub
drosophila_genome <- query(ah, c("2bit", "Drosophila melanogaster", "BDGP6", "dna_sm")) #query for genome sequence
drosophila_genome #select genome sequence
genome <- ah[["AH49674"]]
```

```{r,eval=FALSE}
# we create a new directory that will contain the genome index
dir.create("BDGP6_genome")
# we write the genome sequence in fasta format
export(import.2bit(genome), "BDGP6_genome/genome.fasta.gz", compress=TRUE)
# we build a Rsubread index
Rsubread::buildindex("BDGP6_genome/rsubread", reference="BDGP6_genome/genome.fasta.gz")
```

### Alignment

```{r}
dir.create("aligned")
align.stats <- Rsubread::align(index="BDGP6_genome/rsubread", type="dna",
                               readfile1= "rfastp.trimmed/ENCFF127RRR_R1.fastq.gz",
                               output_file=c("aligned/ENCFF127RRR.bam"),
                               nthreads=6, sortReadsByCoordinates=TRUE)
align.stats
mapped_reads <- align.stats[2,]
total_reads <- align.stats[1,]
mapped_percentage <- (mapped_reads/total_reads)*100
```

# Peak calling

## Using R

```{r}
peaks <- callPeaks("aligned/ENCFF127RRR.bam", fragLength=50L)
num_peaks <- length(peaks)
```


```{r}
#look at peak
head(peaks)
plotSignalTracks("aligned/ENCFF127RRR.bam", region=peaks[1],
                 extend=1000)
#extend: look at 1000 nucleotides next to peak
#since it's a GRanges I can just subset and look at one: peaks[1]
```

#Report

```{r}
report <- paste("Number of mapped reads:", mapped_reads,
                "\nPercentage of mapped reads:", mapped_percentage,
                "\nNumber of peaks found:", num_peaks)
cat(report)

#Plotting of peak inside a gene
q_genome <- query(ah, c("Drosophila", "EnsDb"))
q_genome

drosophila_ensdb <- ah[["AH116255"]]
g <- genes(drosophila_ensdb)
g

#look for peak inside first gene
#define region located on the gene
region <- GRanges("2L", 
              IRanges(start=start(g)[1],
                      end=start(g)[1]+10))
region

#plot the peak inside the defined region
plotSignalTracks("aligned/ENCFF127RRR.bam", region=region,
                 extend=1000, tracks.params=list(ylim=c(0,10)))


```





