---
title: "assignment.rmd"
author: "Robin Scheuplein"
date: "2024-05-02"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(motifmatchr)
  library(MotifDb)
  library(universalmotif)
  library(ggplot2)
  library(SummarizedExperiment) # data structure
  library(sechm) # for plotting heatmaps from a SummrizedExperiment
  library(BiocParallel) # for multithreading
  library(chromVAR) # for motif accessibility estimation
  library(AnnotationHub)
  library(ensembldb)
  library(limma) # for statistical analysis
})
# to control multithreading, unix users can use:
register(MulticoreParam(4))
```

## Download the data

```{r, eval=FALSE}
options(timeout=6000)
download.file("https://ethz-ins.org/content/mouse_mm38_hippocampus.peakCounts.SE.rds", "mouse_mm38_hippocampus.peakCounts", mode="wb")
```

#get genome sequence
```{r}
ah <- AnnotationHub()
genome_search <- AnnotationHub::query(ah, c("Mus musculus", "EnsDb"))
genome_search
genome <- ah[["AH68356"]]

# we'll load it into memory:
genome_seqs <- import(genome)
genome_seqs
Biostrings::writeXStringSet(genome_seqs, "genome.fa")
```

## Get the database of motifs

```{r, eval=FALSE}
motifs <- query(MotifDb, c("HOCOMOCOv10", "Mmusculus")) #specify for hocomoco to have a single set of motifs
motifs
# convert to a format motifmatchr can use, and use the gene symbols as names
motifs <- do.call(TFBSTools::PWMatrixList, setNames(
           universalmotif::convert_motifs(motifs, class="TFBSTools-PWMatrix"),
           mcols(motifs)$geneSymbol))#set names to give the motifs the names of the gene symbol
```

```{r}
# preparing the genome sequence file (here just using a subset)
genome.fa <- Rsamtools::FaFile("genome.fa")
```

1. get the fragment counts for each peak in each sample
  --> add GC bias
2. get the occurence of each motif in each peak 
3. use 1 and 2 to obtain aggregated motif score per sample

```{r}
# get the peaks
peak <- readRDS("mouse_mm38_hippocampus.peakCounts")
# we need to resize the peaks so that they all have a comparable size
peak <- resize(peak,300,fix="center") #fix=center to cut off the peaks at both edges
```

We had the GC concentration in each peak:

```{r}
se <- chromVAR::addGCBias(peak, genome=genome.fa) #to add a column with the GC content in each peak, which will be used as background noise 
#fragment counts for each peak in each sample; add GC bias
rowData(se) 
colnames(se)
```

2. get the occurence of each motif in each peak 

```{r}
# we find which peaks contain which motifs
# we don't use `out="positions"` here, as we just ask a yes or no question; does the peak contain the motif, yes or no?
moi <- motifmatchr::matchMotifs(motifs, subject=se, genome=genome.fa)
assay(se) <- as.matrix(assay(se))
#ensure reproducability as each time you run bg different peaks are used to calculate background noise
set.seed(1234)
# for each peak, we identify similar peaks as background
bg <- chromVAR::getBackgroundPeaks(se, niterations=1000)
# for each motif, we computed per-sample deviations relative to the background
dev <- chromVAR::computeDeviations(object = se, annotations=moi,
                                   background_peaks=bg) #each of the rows is a motif -> relative score for acessability
dev #percentage how accessible each region is
colData(dev)
#check all treatment variables in experiment, i.e. name of different samples
colnames(se)
head(assay(se))
#add column treatment condition
colData(dev)$condition <- c("CTRL", "CTRL", "CTRL", "CTRL", "CTRL", "CTRL", "FSS", "FSS", "FSS", "FSS", "FSS", "FSS")
#add column sex
colData(dev)$sex <- c("Female", "Female", "Female", "Male", "Male", "Male")
#check if columns were added to table
colData(dev) 
```

```{r}
head(assays(dev)$z)
```

# Differential analysis
##linear models: advantage as it iis easier to introduce covariance; group one is at x=0 (mean of group of is interception point at y), and group 2 at x=1; beta1 shows slope that exist between group 1 and group2;p value shows how significantly different the slope from 0 is (meaning that the slope has no gradient and both groups are not different)
```{r}
dev$condition
# if needed, we can specify the baseline condition, by default it's the first in alphabetical order
dev$condition <- factor(dev$condition)
dev$condition <- relevel(dev$condition, "CTRL") #to set the control as primary base level
mm_treatment <- model.matrix(~dev$condition)
# equivalent:
mm_treatment <- model.matrix(~condition, data=as.data.frame(colData(dev)))
#mm <- model.matrix(~sex+dev$~condition, data=as.data.frame(colData(dev)))
mm_treatment

dev$sex
# if needed, we can specify the baseline condition, by default it's the first in alphabetical order
dev$sex <- factor(dev$sex)
dev$sex <- relevel(dev$sex, "Male") #to set the control as primary base level
mm_sex <- model.matrix(~dev$sex)
# equivalent:
mm_sex <- model.matrix(~sex, data=as.data.frame(colData(dev)))
#mm <- model.matrix(~sex+dev$~condition, data=as.data.frame(colData(dev)))
mm_sex
```

#nuclear recetpors have very similar motifs and are difficult to distinguish

```{r}
#comparing stressed (denoted ‘FSS’ – forced swim stress) and control animals
fit_treatment <- limma::eBayes(limma::lmFit(assays(dev)$z, mm_treatment))
res_treatment <- as.data.frame(topTable(fit_treatment, coef="conditionFSS", number = Inf)) #is dropping beta 1 leading to a worse fit of the data #specifiy coef= based on your question e.g. "groupstressed", or "sexmale" "name coefficient according to name set preiously
head(res_treatment)
dim(res_treatment)
res_treatment$TF <- row.names(res_treatment)
ggplot(res_treatment, aes(logFC, -log10(adj.P.Val), label=TF)) + geom_text() 
sechm(dev, features = head(row.names(res_treatment)), assayName="z", top_annotation = c("condition", "depth"))
view_motifs(motifs[c("GCR","PRGR")])
```

```{r}
#comparing male and female animals
fit_sex <- limma::eBayes(limma::lmFit(assays(dev)$z, mm_sex))
res_sex <- as.data.frame(topTable(fit_sex, coef="sexFemale", number = Inf)) #is dropping beta 1 leading to a worse fit of the data #specifiy coef= based on your question e.g. "groupstressed", or "sexmale" "name coefficient according to name set preiously
head(res_sex)
dim(res_sex)
(res_sex)
res_sex$TF <- row.names(res_sex)
ggplot(res_sex, aes(logFC, -log10(adj.P.Val), label=TF)) + geom_text() 
sechm(dev, features = head(row.names(res_sex)), assayName="z", top_annotation = c("condition", "depth", "sex"))
view_motifs(motifs[c("ZFP42","TYY1", "ZN384", "THAP1")])
```


For each analysis, report the top most significant motifs, plot a heatmap of the normalized accessibility scores across the samples for those motifs, and write a short paragraph interpreting the results.
#Report for comparing stressed (denoted ‘FSS’ – forced swim stress) and control animals
The top most significant motifs were GCR and PRGR. Both motifs showed increased flanking accessibility and footprint depth, i.e. the TF activity of both of them was highly upregulated in FSS samples. Noteably, both TF motifs are very similar, thus only one TF might be upregulated.

#Report for comparing male and female animals
The top most significant motifs were ZFP42, TYY1, ZN384, and THAP1. All motifs showed increased accessibility sites in females across the control and FSS condition, suggesting these TFs are not involved in the stress response. Noteably, ZFP42 and TYY1 motifs are very similar, thus only one TF might be upregulated.

