---
title: "assignment.rmd"
author: "Robin Scheuplein"
date: "2024-04-19"
output: html_document
---

```{r}
suppressPackageStartupMessages({
  library(epiwraps)
  library(AnnotationHub)
  library(MotifDb)
  library(memes)
  library(universalmotif)
  library(ensembldb)
  library(ggplot2)
})

ah <- AnnotationHub(localHub=TRUE)
ensdb <- ah[["AH89211"]] # mouse ensembldb object
```


```{r, eval=FALSE}
dir.create("raw") #create new folder
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam", "./raw/atac.chr19.bam")
download.file("https://ethz-ins.org/content/w7/atac.chr19.bam.bai", "./raw/atac.chr19.bam.bai")
```

#preparing tracks
```{r}
##bam files contains all data
bam <- "./raw/atac.chr19.bam"

# create a track using only nucleosome-free fragments, the number of cuts/insertion sites at each position
bam2bw(bam, output_bw = "NF_cuts.bw", paired=TRUE, binWidth=1L, type="ends", extend=2L, minFragLength=30, 
       maxFragLength=120, shift=c(4L,-5L), forceSeqlevelsStyle = "Ensembl")

# create a track using only the (10bp) centers of mono-nucleosome fragments
bam2bw(bam, output_bw = "mono_centers.bw", paired=TRUE, binWidth=5L, minFragLength=140, shift=c(4L,-5L), 
       maxFragLength=220, type="center", extend = 10L, forceSeqlevelsStyle = "Ensembl")


```

## Obtaining the sites with a KLF4 motif

```{r}
# get KLF4 motif
motif <- MotifDb::query(MotifDb, c("Klf4","Mus"))[[3]]
view_motifs(motif) #to check whether it is a high-confidence motif 
motif2 <- convert_motifs(motif, class="TFBSTools-PFMatrix")
genome_search <- AnnotationHub::query(ah, c("Mus musculus"))
genome <- ah[["AH68356"]]

# get the sequence for chr19:
chr19 <- import(genome)["19"]

# find motif matches across chr19
moi_KLF4 <- motifmatchr::matchMotifs(motif2, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_KLF4 <- as(setNames(moi_KLF4,names(chr19)), "GRanges")
```

## Plotting signal around the motif occurences

```{r, fig.width=8, fig.height=4}
# we prepare the list of tracks
tracks <- c("NF cuts"="NF_cuts.bw","Mono centers"="mono_centers.bw")

# extract signals around the motif occurences
# we zoom in to 300bp around the motif centers, in windows of 5bp
sm_KLF4 <- signal2Matrix(tracks, moi_KLF4, w=5, extend=300)

#background normalization
nf <- getNormFactors(tracks, useSeqLevels="19", nwind=5000L)
# then we apply the normalization factors:
sm_KLF4 <- renormalizeSignalMatrices(sm_KLF4, scaleFactors = nf)
plotEnrichedHeatmaps(sm_KLF4, trim=0.95, minRowVal = 12, colors = c("white","darkred"))
```
## Obtaining the sites with a p50 motif

```{r}
# get p50 motif
motif3 <- MotifDb::query(MotifDb, c("p50","Mus"))[[1]]
motif3
view_motifs(motif3) #to check whether it is a high-confidence motif
motif4 <- convert_motifs(motif3, class="TFBSTools-PFMatrix")

# find motif matches across chr19
moi_p50 <- motifmatchr::matchMotifs(motif4, chr19, out="positions", p.cutoff=1e-5)[[1]]
# convert to GRanges
moi_p50 <- as(setNames(moi_p50,names(chr19)), "GRanges")
```


```{r, fig.width=8, fig.height=4}
# we prepare the list of tracks
tracks <- c("NF cuts"="NF_cuts.bw","Mono centers"="mono_centers.bw")

# extract signals around the motif occurences
# we zoom in to 300bp around the motif centers, in windows of 5bp
sm_p50 <- signal2Matrix(tracks, moi_p50, w=5, extend=300)

#background normalization
nf <- getNormFactors(tracks, useSeqLevels="19", nwind=5000L)
# then we apply the normalization factors:
sm_p50 <- renormalizeSignalMatrices(sm_p50, scaleFactors = nf)
plotEnrichedHeatmaps(sm_p50, trim=0.95, colors = c("white","darkred"))
```


